packages/lit-element/src/lib/platform-support.ts

```diff
@@ -302,87 +318,105 @@ interface TemplateStandIn {
-        }
-        this.__baseSetValue(value);
-
-        // Ensure styles are scoped if a TemplateResult is *not* being rendered.
-        // Note, if a TemplateResult *is* being rendered, scoping is done
-        // in `Template._createElement`, before Template parts are created.
-        if (renderingIntoShadowRoot && !renderingTemplateResult) {
-          ensureStyles(currentScope, document.createElement('template'));
-        }
-        currentScope = previousScope;
-        // If necessary move the rendered DOM into the real container.
-        if (renderContainer !== undefined) {
-          // Note, this is the temporary startNode.
-          renderContainer.removeChild(renderContainer.lastChild!);
-          container.insertBefore(
-            renderContainer,
-            this.options?.renderBefore || null
-          );
-          // Un-swizzle tracked nodes.
-          this._startNode = startNode;
-          this._endNode = endNode;
-        }
+  /**
+   * Patch to apply gathered css via ShadyCSS. This is done only once per scope.
+   */
+  NodePart.prototype.__baseSetValue = NodePart.prototype._setValue;
+  NodePart.prototype._setValue = function (
+    this: PatchableNodePart,
+    value: unknown
+  ) {
+    const container = this._startNode.parentNode!;
+    const renderingIntoShadowRoot = container instanceof ShadowRoot;
+    if (!renderingIntoShadowRoot) {
+      this.__baseSetValue(value);
+    } else {
+      const previousScope = currentScope;
+      currentScope = this.options.scope ?? '';
+
+      const renderingTemplateResult = !!(value as ShadyTemplateResult)
+        ?._$litType$;
+      // If result is a TemplateResult decorate it so that when the template
+      // is created, styling can be shimmed.
+      if (renderingTemplateResult) {
+        (value as ShadyTemplateResult).__renderIntoShadowRoot = renderingIntoShadowRoot;
       }
-    },
 
-    /**
-     * Patch NodePart._getTemplate to look up templates in a cache bucketed
-     * by element name.
-     */
-    _getTemplate(strings: TemplateStringsArray, result: ShadyTemplateResult) {
-      let templateCache = scopedTemplateCache.get(currentScope);
-      if (templateCache === undefined) {
-        scopedTemplateCache.set(currentScope, (templateCache = new Map()));
+      // Note, @apply requires outer => inner scope rendering on initial
+      // scope renders to apply property values correctly. Style preparation
+      // is tied to rendering into `shadowRoot`'s and this is typically done by
+      // custom elements. If this is done in `connectedCallback`, as is typical,
+      // the code below ensures the right order since content is rendered
+      // into a fragment first so the hosting element can prepare styles first.
+      // If rendering is done in the constructor, this won't work, but that's
+      // not supported in ShadyDOM anyway.
+      let renderContainer: DocumentFragment | undefined = undefined;
```

Comment:
Ok, so the _key_ thing is "prevent custom elements in the templates I've just rendered from upgrading until I have had a chance to prepare my styles, by not allowing them to connect yet", right?

If so, why doesn't the fact that any elements that might be connected sync to rendering wait a microtask to do _their_ rendering accomplish the same thing?

Reply:
Ah, it's compatibility with legacy Polymer elements that render sync to connected, is that it?